### PR TITLE
end tunnel transition abuse on hard

### PR DESCRIPTION
**Pull request by white-haired-uncle**
I'm sure some little stuff slipped in here, but two main changes (on hard):
1. Your units don't heal when you leave the room.  
2.  When you enter a room, the objectives must be completed before you can go back the way you came in.  In the case of storerooms, you must complete the objectives before you leave at all.  No more stepping out and stepping back in with your army brand new any time you run into a little trouble.  You can still escape if you can make it to a different exit (for now, long term on hard you should probably not have a way out).  Also, every room now has objectives.

I changed the vars, like sc08=1 and sc09=1 to ch5_quests.sc08.done=yes and ch5_quests.sc09.done=yes.  I find this to help a lot when troubleshooting as if they are all under one variable the inspect gui show you them all at once instead of making you click all over, and you can copy to clipboard.  Also, putting putting .done and .undead under ch5_quests.sc22 groups them nicely.  Haven't done items yet.  There's a converter added to prestart which I tested on an old save as best I could.

### CODE REVIEW

**Review by Dugy — CHANGES_REQUESTED** [review]
I don't quite get a few things

File: scenarios5/17_Amplificator3.cfg [hunk 1]
```diff
@@ -58,6 +58,43 @@
...
+        [if]
+            [variable]
+                name=ch5_quests.sc17.done
+                equals=yes
+            [/variable]
+            [then]
+                {TRANSITION 20-27,0-4 23,1 _"Amplificator 2" 15_Amplificator_2}
+                {TRANSITION 50-54,6-8 52,7 _"Southeastern Storeroom" 16_Southeastern_Storeroom}
+                {TRANSITION 1-3,14-19 1,17 _"Soldiers' Training Room" 19_Soldiers_Training_Room}
+            [/then]
```
[the remark this answers is not in the document]
Comment (white-haired-uncle) [reply scenarios5/17_Amplificator3.cfg:71]: I think you looked only at the if/then part.  In the else, it opens transitions, just not for the way you came in.

File: scenarios5/08_Entrance.cfg [hunk 2]
```diff
@@ -111,12 +133,31 @@
...
+                    message= _ "I suggest we do just that, people so scared rarely lie."
                 [/message]
                 [message]
                     speaker=Lethalia
                     message= _ "Fortunately it is so dark here. Our undead will have an advantage."
                 [/message]
+                [message]
+                    speaker=Efraim
+                    message= _ "Good point.  Perhaps we should have brought more of them."
+                    [show_if] # No idea what the actual limit we should use here is.  Also checking Valahlla left as an exercise for the reader.
```
[the remark this answers is not in the document]
Comment (white-haired-uncle) [reply scenarios5/08_Entrance.cfg:145]: No, Valahlla left as an exercise for the reader.  This checks the autorecalled units and the recall list.

[the remark this answers is not in the document]
Comment (Dugy) [reply scenarios5/08_Entrance.cfg:145]: Ahaaa.

File: scenarios5/17_Amplificator3.cfg (hunk 1 above)
[the remark this answers is not in the document]
Comment (Dugy) [reply scenarios5/17_Amplificator3.cfg:71]: Aha, yeah.

**Comment by Dugy** [comment]
Sorry for the confusion.

**Comment by white-haired-uncle** [comment]
I'm working on amplificators now.  I'm working on a new unit, the suicide drone.  I want the amplificator to give them a movement boost when they start next to him.  It's hard to troubleshoot, as I can't see the beginning of his turn.  You see anything wrong with this?

```
    [event]
        name=side turn
        id=propel
        first_time_only=no
        [modify_unit]
            [filter]
                side=$side_number
                [filter_adjacent]
                    ability=propel
                    is_enemy=no
                [/filter_adjacent]
            [/filter]
            moves=$($unit.moves + 2)
        [/modify_unit]
    [/event]

```

I swear there's some ability that does this already (though +1MP) that I could copy if I could find it.  Right now I'm just trying to copy off Duke's push.

**Pull request merged**